File: src/main/java/com/windchopper/fs/sftp/SftpFileSystem.java
package com.windchopper.fs.sftp;

import com.github.windchopper.common.util.Pipeliner;
import com.jcraft.jsch.ChannelSftp;
import com.jcraft.jsch.ChannelSftp.LsEntry;
import com.jcraft.jsch.JSchException;
import com.jcraft.jsch.Session;
import com.jcraft.jsch.SftpException;
import org.apache.commons.collections4.map.LRUMap;
import org.apache.commons.lang3.StringUtils;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.UserPrincipalLookupService;
import java.util.*;
import java.util.stream.Stream;

import static java.util.Arrays.stream;
import static java.util.Collections.emptySet;
import static java.util.stream.Collectors.toList;

public class SftpFileSystem extends FileSystem implements SftpConstants {

    private final SftpFileSystemProvider provider;
    private final SftpConfiguration configuration;

    private final Map<String, SftpFile> viewBuffer;
    private final Map<String, List<SftpFile>> listBuffer;

    private final Session session;

    SftpFileSystem(SftpFileSystemProvider provider, SftpConfiguration configuration) throws IOException {
        this.provider = provider;
        this.configuration = configuration;

        try {
            viewBuffer = new LRUMap<>(configuration.bufferSize);
            listBuffer = new LRUMap<>(configuration.bufferSize);

            session = Pipeliner.of(provider.jsch)
                .mapFailable(configuration.sessionIdentity::createJschSession)
                .accept(session -> session.setConfig("StrictHostKeyChecking", "no"))
                .set(session -> session::setPassword, configuration.password)
                .acceptFailable(Session::connect)
                .get();
        } catch (JSchException thrown) {
            throw new IOException(thrown);
        }
    }

    SftpFile view(String path) throws IOException {
        SftpFile file = viewBuffer.get(path);

        if (file == null) {
            try {
                ChannelSftp channel = (ChannelSftp) session.openChannel("sftp");
                channel.connect();

                try {
                    fillInBuffers(path, channel.ls(path));
                    file = viewBuffer.get(path);
                } finally {
                    channel.disconnect();
                }
            } catch (JSchException | SftpException thrown) {
                throw new IOException(thrown);
            }
        }

        if (file == null) {
            throw new FileNotFoundException(path);
        }

        return file;
    }

    List<SftpFile> fillInBuffers(String path, Vector<?> entries) {
        List<SftpFile> files = new ArrayList<>(entries.size());

        for (Object entry : entries) {
            LsEntry typedEntry = (LsEntry) entry;
            SftpFile file;

            if (typedEntry.getFilename().equals("..")) {
                continue;
            }

            if (typedEntry.getFilename().equals(".")) {
                file = new SftpFile(StringUtils.substringBeforeLast(path, SEPARATOR), StringUtils.substringAfterLast(path, SEPARATOR), typedEntry.getAttrs());
            } else {
                files.add(file = new SftpFile(path, typedEntry.getFilename(), typedEntry.getAttrs()));
            }

            viewBuffer.put(file.toAbsolutePath(), file);
        }

        listBuffer.put(path, files);
        return files;
    }

    List<SftpFile> list(String path) throws IOException {
        List<SftpFile> files = listBuffer.get(path = path.equals(SEPARATOR) ? path : StringUtils.removeEnd(path, SEPARATOR));

        if (files == null) {
            try {
                ChannelSftp channel = (ChannelSftp) session.openChannel("sftp");
                channel.connect();

                try {
                    files = fillInBuffers(path, channel.ls(path));
                } finally {
                    channel.disconnect();
                }
            } catch (JSchException | SftpException thrown) {
                throw new IOException(thrown);
            }
        }

        return files;
    }

    @Override public SftpFileSystemProvider provider() {
        return provider;
    }

    @Override public void close() {
        session.disconnect();
        provider.retire(configuration.sessionIdentity, this);
        List.of(viewBuffer, listBuffer)
            .forEach(Map::clear);
    }

    DirectoryStream<Path> newDirectoryStream(SftpPath path, DirectoryStream.Filter<? super Path> filter) throws IOException {
        return new SftpDirectoryStream(list(path.toString()).stream()
            .map(file -> file.toPath(this, configuration.sessionIdentity))
            .collect(toList()));
    }

    @Override public boolean isOpen() {
        return session.isConnected();
    }

    @Override public boolean isReadOnly() {
        return false;
    }

    @Override public String getSeparator() {
        return SftpConfiguration.SEPARATOR;
    }

    String realPath(String path) throws IOException {
        try {
            ChannelSftp channelSftp = (ChannelSftp) session.openChannel("sftp");
            channelSftp.connect();

            try {
                return channelSftp.realpath(path);
            } finally {
                channelSftp.disconnect();
            }
        } catch (JSchException | SftpException thrown) {
            throw new IOException(
                String.format("Couldn't resolve path %s", path),
                thrown);
        }
    }

    @Override public Iterable<Path> getRootDirectories() {
        return List.of(new SftpPath(this, configuration.sessionIdentity, SftpConfiguration.SEPARATOR));
    }

    @Override public Iterable<FileStore> getFileStores() {
        throw new UnsupportedOperationException();
    }

    @Override public Set<String> supportedFileAttributeViews() {
        return emptySet();
    }

    @Override public Path getPath(String firstPathElement, String... restPathElements) {
        return new SftpPath(this, configuration.sessionIdentity, Stream.concat(
            Stream.of(firstPathElement),
            stream(restPathElements))
                .toArray(String[]::new));
    }

    @Override public PathMatcher getPathMatcher(String path) {
        throw new UnsupportedOperationException();
    }

    @Override public UserPrincipalLookupService getUserPrincipalLookupService() {
        throw new UnsupportedOperationException();
    }

    @Override public WatchService newWatchService() {
        throw new UnsupportedOperationException("Couldn't watch remote file system");
    }

}

File: src/main/kotlin/com/github/windchopper/fs/sftp/SftpFileSystem.kt
package com.github.windchopper.fs.sftp

import com.github.windchopper.fs.internal.jsch.JSchHelper
import com.github.windchopper.fs.internal.rethrow
import com.jcraft.jsch.ChannelSftp.LsEntry
import org.apache.commons.collections4.map.LRUMap
import java.io.FileNotFoundException
import java.io.IOException
import java.nio.channels.SeekableByteChannel
import java.nio.file.*
import java.nio.file.attribute.UserPrincipalLookupService
import java.util.*

class SftpFileSystem(private val provider: SftpFileSystemProvider, private val configuration: SftpConfiguration): FileSystem() {

    companion object {

        const val SCHEME = "sftp"
        const val PATH_SEPARATOR = "/"

    }

    val viewBuffer: MutableMap<String, SftpFile> = LRUMap()
    val listBuffer: MutableMap<String, List<SftpFile>> = LRUMap()

    val helper = JSchHelper(JSchHelper.Type.SFTP, configuration.channelInactivityDuration, rethrow(::IOException) {
            with (configuration) {
                with (sessionIdentity) {
                    provider.secureChannel.getSession(username, host, port)
                        .let {
                            it.setConfig("StrictHostKeyChecking", "no")
                            it.setPassword(password)
                            it.connect()
                            it
                        }
                }
            }
        })

    fun view(pathString: String): SftpFile {
        var bufferedView = viewBuffer[pathString]

        if (bufferedView == null) bufferedView = helper.performConnected { channel ->
            fillInBuffers(pathString, channel.ls(pathString))
            viewBuffer[pathString]
        }

        return bufferedView
            ?:throw FileNotFoundException(pathString)
    }

    @Suppress("UNCHECKED_CAST") fun fillInBuffers(pathString: String, entries: Vector<*>): List<SftpFile> {
        val files: MutableList<SftpFile> = ArrayList(entries.size)

        (entries as Vector<LsEntry>)
            .filter {
                it.filename != ".."
            }
            .map { entry ->
                if (entry.filename == ".") {
                    SftpFile(pathString.substringBeforeLast(PATH_SEPARATOR), pathString.substringAfterLast(PATH_SEPARATOR), entry.attrs)
                } else {
                    SftpFile(pathString, entry.filename, entry.attrs)
                        .let { file ->
                            files.add(file)
                            file
                        }
                }
            }
            .forEach { file ->
                with (file.toAbsolutePath()) {
                    viewBuffer[this] = file
                }
            }

        listBuffer[pathString] = files

        return files
    }

    @Throws(IOException::class) fun list(path: String): List<SftpFile> {
        return (if (path == PATH_SEPARATOR) path else path.removeSuffix(PATH_SEPARATOR))
            .let {
                listBuffer[it]
                    ?:helper.performConnected { channel ->
                        fillInBuffers(path, channel.ls(path))
                    }
            }
    }

    override fun provider(): SftpFileSystemProvider {
        return provider
    }

    @Throws(IOException::class) override fun close() {
        helper.session.disconnect()
        provider.retire(configuration.sessionIdentity, this)
        for (buffer in listOf(viewBuffer, listBuffer)) {
            buffer.clear()
        }
    }

    fun newDirectoryStream(path: Path, filter: DirectoryStream.Filter<in Path?>): DirectoryStream<Path> {
        return SftpDirectoryStream(filter, list(path.toString())
            .map { it.toPath(this, configuration.sessionIdentity) })
    }

    fun newByteChannel(path: Path): SeekableByteChannel {
        return helper.performConnected { channel ->
            // todo SftpFileChannel(helper, path)
            // full download for now

            val tempFilePath = Files.createTempFile("sftp-", "-download")

            Files.newOutputStream(tempFilePath).use { outputStream ->
                channel.get(path.toAbsolutePath().toString(), outputStream)
            }

            Files.newByteChannel(tempFilePath, StandardOpenOption.DELETE_ON_CLOSE)
        }
    }

    override fun isOpen(): Boolean {
        return helper.session.isConnected
    }

    override fun isReadOnly(): Boolean {
        return false
    }

    override fun getSeparator(): String {
        return PATH_SEPARATOR
    }

    @Throws(IOException::class) fun realPath(path: String?): String {
        return helper.performConnected {
            it.realpath(path)
        }
    }

    override fun getRootDirectories(): Iterable<Path> {
        return listOf(SftpPath(this, configuration.sessionIdentity, PATH_SEPARATOR))
    }

    override fun getFileStores(): Iterable<FileStore> {
        TODO("not implemented")
    }

    override fun supportedFileAttributeViews(): Set<String> {
        return emptySet()
    }

    override fun getPath(firstPathElement: String, vararg restPathElements: String): Path {
        return SftpPath(this, configuration.sessionIdentity, *arrayOf(firstPathElement).plus(restPathElements))
    }

    override fun getPathMatcher(path: String): PathMatcher {
        TODO("not implemented")
    }

    override fun getUserPrincipalLookupService(): UserPrincipalLookupService {
        TODO("not implemented")
    }

    override fun newWatchService(): WatchService {
        throw UnsupportedOperationException("Couldn't watch remote file system")
    }

}